src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,173 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "math/vector/Int.hpp"
+#include "math/vector/Float.hpp"
+#include "math/vector/Size_t.hpp"
+#include "math/vector/math_functor/abs.hpp"
+#include "math/vector/math_functor/max.hpp"
+#include "cuSTL/container/PseudoBuffer.hpp"
+#include "dataManagement/DataConnector.hpp"
+#include "fields/FieldJ.hpp"
+#include "math/Vector.hpp"
+#include "cuSTL/algorithm/mpi/Gather.hpp"
+#include "cuSTL/algorithm/mpi/Reduce.hpp"
+#include "cuSTL/container/DeviceBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/algorithm/kernel/Foreach.hpp"
+#include "cuSTL/algorithm/host/Foreach.hpp"
+#include "lambda/Expression.hpp"
+#include <sstream>
+
+#include "cuSTL/algorithm/kernel/Reduce.hpp"
+
+namespace picongpu
+{
+
+template<typename Species>
+ChargeConservation<Species>::ChargeConservation()
+    : name("print the maximum charge derivation between particles and div E"), 
+      prefix("electronChargeConservation")
+{
+    Environment<>::get().PluginConnector().registerPlugin(this);
+}
+
+template<typename Species>
+void ChargeConservation<Species>::pluginRegisterHelp(po::options_description& desc)
+{
+    desc.add_options()
+        ((this->prefix + "_frequency").c_str(),
+        po::value<uint32_t > (&this->notifyFrequency)->default_value(0), "notifyFrequency");
+}
+
+template<typename Species>
+std::string ChargeConservation<Species>::pluginGetName() const {return this->name;}
+
+template<typename Species>
+void ChargeConservation<Species>::pluginLoad()
+{
+    Environment<>::get().PluginConnector().setNotificationPeriod(this, this->notifyFrequency);
+}
+
+template<int dim, typename ValueType>
+struct Div;
+
+template<typename ValueType>
+struct Div<DIM3, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        const ValueType reciDepth = float_X(1.0) / CELL_DEPTH;
+        return ((*field(1,0,0)).x() - (*field).x()) * reciWidth +
+               ((*field(0,1,0)).y() - (*field).y()) * reciHeight +
+               ((*field(0,0,1)).z() - (*field).z()) * reciDepth;
+    }
+};
+
+template<typename ValueType>
+struct Div<DIM2, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        return ((*field(1,0)).x() - (*field).x()) * reciWidth +
+               ((*field(0,1)).y() - (*field).y()) * reciHeight;
+    }
+};
+
+template<typename Species>
```

Comment:
remove template, should work for all species (add results in FieldTmp) with patch below